onnx/reference/ops/op_string_concat.py

```diff
@@ -0,0 +1,23 @@
+# Copyright (c) ONNX Project Contributors
+
+# SPDX-License-Identifier: Apache-2.0
+# pylint: disable=R0912,R0913,W0221
+
+import numpy as np
+
+from onnx.reference.op_run import OpRun
+
+_acceptable_str_dtypes = ("U", "O")
+
+
+class StringConcat(OpRun):
+    def _run(self, x, y):
+        if (
+            x.dtype.kind not in _acceptable_str_dtypes
+            or y.dtype.kind not in _acceptable_str_dtypes
+        ):
+            raise TypeError(
+                f"Inputs must be string tensors, received dtype {x.dtype} and {y.dtype}"
+            )
+        # As per onnx/mapping.py, object numpy dtype corresponds to TensorProto.STRING
```

Comment:
This part (as well as the _acceptable_str_dtypes above, which seems related) is a bit unclear to me. Perhaps @xadupre or @jcwchen can articulate the difference. Why object type over string type? It would help for the reference-implementation to support just one specific representation type.

Reply:
Ok, I see that we need to use object type to get variable length strings. The question is do we need to support inputs that are not object-type (that is, unicode I guess above). I can see it may be convenient, but it may also potentially be helpful to rigidly enforce expected representation types. Just wondering.

Reply:
> Why object type over string type?

I think just using `np.str_` makes more sense to me however ONNX currently relates `TensorProto.STRING` to `np.object` quite explicitly.

> The question is do we need to support inputs that are not object-type (that is, unicode I guess above).

Slightly confused by this.
```
>>> import numpy as np
>>> np.array(["的", "中"])
array(['的', '中'], dtype='<U1')
>>> np.issubdtype(np.array(["的", "中"]).dtype, np.str_)
True
```